### PR TITLE
Cpp implementation

### PR DESCRIPTION
**Pull request by esteve**
This is just a quick and rather hacky C++ port of the Javascript translator. I've added a dependency on Hunter and `fmtlib` for the string manipulation code and also copied over the `.clang-format` from cpp-ethereum.

Connects to #4

### CODE REVIEW

**Comment by esteve** [comment]
I think Travis is failing because the C++ compiler/standard library is rather old, there doesn't seem to be anything wrong with that piece of code.

**Comment by lrettig** [comment]
Thanks @esteve! @chfast you're probably the right person to review this.

**Comment by axic** [comment]
Is `wast-sync.h` and `wast-async.h` handwritten, because I don't see any generator tool?

File: bin/evm2wasm.js [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env node
+#!/usr/bin/env nodejs
```
Comment (axic) [inline bin/evm2wasm.js]: Why is this needed?

File: libs/evm2wasm/evm2wasm.cpp [hunk 2]
```diff
@@ -1,61 +1,761 @@
...
+}
+
+string evm2wasm(const string& input)
+{
+    return wast2wasm(evm2wast(input));
+}
+
+std::string opcodeToString(opcodeEnum opcode)
+{
+    switch (opcode)
```
Comment (axic) [inline libs/evm2wasm/evm2wasm.cpp]: How about a map for this?

File: include/evm2wasm.h [hunk 3]
```diff
@@ -1,5 +1,305 @@
...
+    memset,
+    memusegas,
+    mod_320,
+    mod_512,
+    mul_256
+};
+
+// maps the async ops to their call back function
+static std::map<opcodeEnum, std::string> callbackFuncs = {{opcodeEnum::SSTORE, "$callback"},
+    {opcodeEnum::SLOAD, "$callback_256"}, {opcodeEnum::CREATE, "$callback_160"},
```
[the remark this answers is not in the document]
Comment (esteve) [reply include/evm2wasm.h]: Fixed in https://github.com/ewasm/evm2wasm/pull/201/commits/ea8ff91825044355f950c374f83dd3a990a891e3

File: bin/evm2wasm.js (hunk 1 above)
Comment (esteve) [reply bin/evm2wasm.js]: This one was a leftover from when @lrettig and I tried the Javascript translator, Node is installed as nodejs in Debian.

Reverted in https://github.com/ewasm/evm2wasm/pull/201/commits/e5199247e992d81c75f98c88650a9b537f71bf48

File: include/evm2wasm.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (chfast) [reply include/evm2wasm.h]: @axic You have to give up hand-tuning the code formatting.

File: libs/evm2wasm/evm2wasm.cpp (hunk 2 above)
Comment (esteve) [reply libs/evm2wasm/evm2wasm.cpp]: I wanted to avoid callers of this function to deal with `std::pair`, but I can change it to a map if you prefer.

File: include/evm2wasm.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (axic) [reply include/evm2wasm.h]: This format is super hard to read, there must be a way to disable it.

**Comment by esteve** [comment]
> Is wast-sync.h and wast-async.h handwritten, because I don't see any generator tool?

Just a bit of `sed` here and there in the `wast.js` and `wast-async.js`, but I'm going to write a generator so it's all done automatically.

> Also I'd prefer them to be .cpp files to avoid recompiling them if not needed 

Yeah, I agree, it's rather awful :smile: I'm going to split them up and see if they look better.

File: include/evm2wasm.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (chfast) [reply include/evm2wasm.h]: I've found a trick to do it in clang-format: add `,` after the last item.

**Comment by chfast** [comment]
I'm not sure how much work it is, but I'd rather stick to C++ library only. Adding additional dependencies is quite a big deal and we also want to compile that to WASM later on. If possible I'd use just `std::string` concatenation.

**Comment by axic** [comment]
@esteve do you want to create a new PR for the samples?

**Comment by axic** [comment]
Also can you please rebase and squash some commits?

**Comment by lrettig** [comment]
The samples are mine, I can pull these out into a separate PR, good idea

**Comment by lrettig** [comment]
But go for it if you get to it first, @esteve!

**Comment by esteve** [comment]
> @axic do you want to create a new PR for the samples?
> @lrettig But go for it if you get to it first, @esteve!

Done! :-) #215

**Comment by axic** [comment]
@lrettig can you rebase this so it runs on circleci? Also the first and last commit are not useful anymore (the useful parts have been merged since).

**Comment by esteve** [comment]
@axic done, sorry for the delay. Been busy lately, but I'll address the pending feedback in the upcoming days (I hope :-))

File: include/evm2wasm.h [hunk 4]
```diff
@@ -351,9 +351,9 @@ struct WastCode
     std::string imports;
 };
 
-std::string evm2wasm(const std::string& input);
+std::string evm2wasm(const std::vector<char>& input);
```
Comment (axic) [inline include/evm2wasm.h]: Why change the public API? And what is the desired content after the change?

Comment (chfast) [reply include/evm2wasm.h]: This is bad choice of the type. If you want to handle multiple input types consider using `string_view` (not in standard yet) or pass iterators.

**Comment by axic** [comment]
Need to update after #263.

**Comment by jwasinger** [comment]
In order to use this branch, use [this](https://github.com/ewasm/hera/pull/237) branch of Hera.  This change corresponds to the changes in b9d6a00f079a3538c76cc745f54e026b100e98a6 which make changes to the evm2wasm public API.

**Comment by axic** [comment]
Pulled out a lot of changes and rebased about 50 times until finally it looks ok 😉

**Comment by axic** [comment]
Once CI is running (depends on some Hera PRs) I'd vote for merging this, even if it is not fully passing tests.

**Comment by chfast** [comment]
I'm ok with this.

**Comment by axic** [comment]
I think all the PUSHes are broken, but at least that fuzzes Hera+testeth:
```
getBalance ffffffe0 ffffffe0
unknown location(0): fatal error: in "GeneralStateTests/stStackTests": memory access violation at address: 0x108f57920: no mapping at fault address
/home/builder/project/test/tools/libtesteth/ImportTest.cpp(257): last checkpoint
```

**Comment by axic** [comment]
Some of the last bugs left:
- #301 which caused missing dependencies (since that list is not expanded recursively, not sure why JS didn't care..)
- and this damn typo of `gt_s` vs `lt_s` in the last commit...
- state test mismatch :(
